File: 200/214 - Shortest Palindrome.java
public class Solution {
    public String shortestPalindrome(String s) {
        StringBuilder sb = new StringBuilder(s);
        String rev = new StringBuilder(s).reverse().toString();
        int i, j;
        int[] next;
        
        sb.append('*');
        sb.append(rev);
        next = new int[sb.length()];
        
        for (i = 1; i < next.length; ++i) {
            j = next[i - 1];
            while (j > 0 && sb.charAt(i) != sb.charAt(j))
                j = next[j - 1];
            
            next[i] = j == 0 ? (sb.charAt(i) == sb.charAt(0) ? 1 : 0) : j + 1;
        }  
        
        return rev.substring(0, rev.length() - next[next.length - 1]) + s;
    }
}

File: csharp/214.shortest-palindrome.cs
/*
 * @lc app=leetcode id=214 lang=csharp
 *
 * [214] Shortest Palindrome
 */

using System;

public class Solution {
    public string ShortestPalindrome(string s) {
        var arr = s.ToCharArray();
        Array.Reverse(arr);
        string rev = new string(arr);
        string t = s + '$' + rev;
        int[] next = new int[t.Length];
        for (int i = 1; i < t.Length; ++i) {
            int j = next[i - 1];
            while (j > 0 && t[i] != t[j]) {
                j = next[j - 1];
            }
            if (t[i] == t[j]) {
                j += 1;
            }
            next[i] = j;
        }
        return rev.Substring(0, rev.Length - next[t.Length - 1]) + s;
    }
}

